### PR TITLE
InfusatesLoader

### LINKED ISSUE #911: Isotope pattern match fails in some cases
### Problem

The pattern used to parse an isotopes string fails with a `ValueError` on some infusates with isotope position data.

### Steps to reproduce

1. Call `parse_infusate_name` on an infusate name like `ti {glucose-[2,3-13C2,4-17O1];C16:0-[5,6-13C2,17O2]}`, with concentrations like `[1.0, 2.0]` (though the concentrations don't matter)

### Current behavior

```
ValueError: invalid literal for int() with base 10: ''
```

I did a debug print of the match and the string.

- This series of isotopes is isolated for individual isotope parsing: `2,3-13C2,4-17O1`
- But this is the match from `finditer`: `match=',4-17O1'`

Note the leading comma.

### Expected behavior

No error and correct resulting infusate data.

### Suggested Change

I think that the solution will involve forcing the comma to split the individual isotope data.

### Comment

This and another bug I discovered are the reasons I haven't submitted the PR I thought I would have submitted yesterday at COB.

-----

### Assumptions

1. List of assumptions made WRT the code
2. E.g. We will assume input is correct (explaining why there is no validation)

### Limitations

1. A list of things this work will specifically not do
2. E.g. This feature will only handle the most frequent use case X

### Affected Components

- change: File path or DB table ...
- add: Environment variable or server setting
- delete: External executable or cron job

### Requirements

- [ ] 1. List of numbered conditions to be met for the feature
- [ ] 2. E.g. Every column/row must display a value, i.e. cannot be empty
- [ ] 3. Numbers for reference & checkboxes for progress tracking

#### GUI Change description

None provided

#### Code Change Description

None provided

#### Tests

- [ ] 1. A description of at least one test for each requirement above.
- [ ] 2. E.g. Test for req 2 that there's an exception when display value is ''
- [ ] 3. Numbers for reference & checkboxes for progress tracking

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

Added an infusates loader and fixed a few bugs that were discovered in the middle and at the end of development and in the way of completion.

See the commit log messages for descriptions of the bug fixes, but I will comment in-line.

My linting script also changed some files I touched (and then reverted).

## Affected Issues/Pull Requests

- Partially addresses #822
- Resolves #911
- Next PR: #915

## Review Notes

There is a bandaid and an issue I will create issues out of, which I will comment on in-line.

## Checklist

This pull request will be merged once the following requirements are met.  The
author and/or reviewers should uncheck any unmet requirements:

- Review requirements
  - Minimum approvals: 1 
  - No changes requested
  - All blocking issues resolved by reviewers
  - Specific reviewers: @__add_username_here__
    
  - Review period: 2 days 
- Associated issue/pull request requirements:
  
  - [x] All requirements in affected issues marked "resolved" are satisfied
  - [x] All required pull requests are merged *(or none)*
- Basic requirements
  
  - [x] [All linters pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All conflicts resolved
- Overhead requirements
  
  - [x] [New/changed method tests implemented/updated *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [ ] [Updated *Unreleased* section of `changelog.md` *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md)
  - [x] [Migrations created & committed *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)

### CODE REVIEW

File: DataRepo/loaders/infusates_loader.py [hunk 1]
```diff
@@ -0,0 +1,966 @@
...
+            GROUP_NAME_KEY,
+            TRACER_NAME_KEY,
+            CONC_KEY,
+        ],
+        [
+            NAME_KEY,
+            GROUP_NAME_KEY,
+            TRACER_NAME_KEY,
+            CONC_KEY,
+        ],
```
Comment (hepcat72) [inline DataRepo/loaders/infusates_loader.py:102]: C. The only difference between these 2 unique constraints is ID versus NAME.  Each must be unique with the other data.

File: DataRepo/loaders/infusates_loader.py [hunk 2]
```diff
@@ -0,0 +1,966 @@
...
+        Infusate.__name__: {
+            "name": NAME_KEY,
+            "tracer_group_name": GROUP_NAME_KEY,
+        },
+    }
+
+    # List of model classes that the loader enters records into.  Used for summarized results & some exception handling
+    Models = [InfusateTracer, Infusate]
+
+    def __init__(self, *args, **kwargs):
```
Comment (hepcat72) [inline DataRepo/loaders/infusates_loader.py:121]: C. Below this point, the implementation is nearly verbatim from the `tracer_loader`, with variable name changes.

File: DataRepo/loaders/infusates_loader.py [hunk 3]
```diff
@@ -0,0 +1,966 @@
...
+                        {
+                            "tracer_name": parsed_infusate["tracer"]["unparsed_string"],
+                            "tracer_concentration": parsed_infusate["concentration"],
+                            # The row info where the tracer name was obtained
+                            "rownum": table_infusate["rownum"],
+                            "row_index": table_infusate["row_index"],
+                        }
+                    )
+
+    def get_or_create_just_infusate(self, infusate_dict):
```
Comment (hepcat72) [inline DataRepo/loaders/infusates_loader.py:563]: C. Note, if the infusate record exists, all the associated records exist.  The search is done using all that data.  If it doesn't exist, only the infusate record is created.  This is different from the get or create in the infusate class.

File: DataRepo/loaders/infusates_loader.py [hunk 4]
```diff
@@ -0,0 +1,966 @@
...
+                    sheet=self.sheet,
+                )
+            )
+            for nk in self.inconsistent_numbers[infusate_name].keys():
+                self.add_skip_row_index(
+                    index_list=self.inconsistent_numbers[infusate_name][nk]
+                )
+
+    def check_infusate_name_consistent(self, rec, infusate_dict):
+        """Checks for consistency between a dynamically generated infusate name and the one supplied in the file.
```
Comment (hepcat72) [inline DataRepo/loaders/infusates_loader.py:876]: - [ ] TODO: Update this docstring to reflect that the concentration data is involved in the consistency check and that the order of tracers and isotopes does not matter (even though the name in the db records order those things).

Comment (hepcat72) [reply DataRepo/loaders/infusates_loader.py:876]: It might be worthwhile to check that any other pre-existing code the searches based on name strings takes these issues into account.

File: DataRepo/loaders/table_loader.py [hunk 5]
```diff
@@ -1955,6 +1956,27 @@ def load_wrapper(self, *args, **kwargs):
...
+                            # exception and remove them.  (Added due to extra exception in
+                            # test_infusate_loader_load_data before that test's original issue was fixed)
+                            # TODO: There probably exists a better way to do this, but there does not exist a rollback
+                            # hook in django's atomic transaction architecture.  What *should* happen is that the
+                            # autoupdate buffer should be cleared upon any exception that causes a rollback.  I have a
+                            # few ideas, such as the fact that the decorators could probably be applied in places that
+                            # would avoid this, but this bandaid will suffice for now.
+                            self.aggregated_errors_object.remove_exception_type(
+                                AutoUpdateFailed
+                            )
```
Comment (hepcat72) [inline DataRepo/loaders/table_loader.py:1978]: C. This is a bandaid.
- [ ] TODO: Create an issue to catch this exception in maintained model and move the defer_autoupdates decorator to a place between the raised exception inside the atomic transaction block and the catch outside the atomic transaction so the buffer can be cleared.

This is newly necessary due to the new individual record rollback strategy, however, it's a *minor* issue.  The only result is an extra exception about this failed autoupdate.  There is an outer atomic transaction block that rolls back everything, even these deferred updates.  The only reason it shows up here is because the tested method doesn't have a defer decorator, so the mass autoupdate is attempted.

File: DataRepo/loaders/tracers_loader.py [hunk 6]
```diff
@@ -28,7 +28,7 @@ class TracersLoader(TableLoader):
     LABEL_POSITIONS_KEY = "LABELPOSITIONS"
     NAME_KEY = "NAME"
 
-    POSITIONS_DELIMITER = ","
+    POSITIONS_DELIMITER = TracerLabel.POSITIONS_DELIMITER
```
Comment (hepcat72) [inline DataRepo/loaders/tracers_loader.py:31]: C. I made these class attributes so I could reference them via the class elsewhere.

File: DataRepo/loaders/tracers_loader.py [hunk 7]
```diff
@@ -175,7 +175,7 @@ def init_load(self):
         Returns:
             Nothing
         """
-        self.tracer_dict = defaultdict(dict)
+        self.tracers_dict = defaultdict(dict)
```
Comment (hepcat72) [inline DataRepo/loaders/tracers_loader.py:178]: C. I did some variable renaming to better distinguish between an individual tracer dict and the dict with all the tracers.

Comment (hepcat72) [reply DataRepo/loaders/tracers_loader.py:178]: All the changes in this file are either that type of rename or a typo/stake comment fix.

File: DataRepo/models/infusate.py [hunk 8]
```diff
@@ -12,12 +13,6 @@
     from DataRepo.utils.infusate_name_parser import InfusateData
 
 
-CONCENTRATION_SIGNIFICANT_FIGURES = 3
-TRACER_DELIMETER = ";"
-TRACERS_LEFT_BRACKET = "{"
-TRACERS_RIGHT_BRACKET = "}"
```
Comment (hepcat72) [inline DataRepo/models/infusate.py:18]: C. Moved into the class for access.

File: DataRepo/models/infusate.py [hunk 9]
```diff
@@ -119,7 +119,7 @@ def __str__(self):
 
     @MaintainedModel.setter(generation=0, update_field_name="name", update_label="name")
     def _name(self):
-        # Format: `tracer_group_name{tracername;tracername}`
+        # Format: `tracer_group_name {tracername[concentration];tracername[concentration]}`
```
Comment (hepcat72) [inline DataRepo/models/infusate.py:122]: C. Better reflection of what this does.

File: DataRepo/models/infusate.py [hunk 10]
```diff
@@ -131,21 +131,131 @@ def _name(self):
...
                     link_recs.all(),
                 )
             )
         )
 
         if self.tracer_group_name is not None:
-            name = f"{self.tracer_group_name} {TRACERS_LEFT_BRACKET}{name}{TRACERS_RIGHT_BRACKET}"
+            name = f"{self.tracer_group_name} {self.TRACERS_LEFT_BRACKET}{name}{self.TRACERS_RIGHT_BRACKET}"
 
         return name
```
Comment (hepcat72) [inline DataRepo/models/infusate.py:148]: C. The following 2 new methods allow me to error-check that the names match the data in the file, and the resulting record created in the database.  I suggest reading the docstring for `infusate_name_equal`.

Note that after implementing these, I noted that they are actually unnecessary after development (they were useful for debugging).  The `get_or_create_infusate`, etc. would always result in **getting** an existing record, regardless of order, etc (though I didn't look at how it handles the float equality issue).

However, having these methods will allow users to search using infusate names regardless of the tracer/isotope order.  It would just have to be integrated.  I'm amenable to removing them.

But, if we do decide to keep them:

- [ ] TODO: Create an issue to break this method up across the other classes (Tracer, InfusateTracer, and TracerLabel)

File: DataRepo/models/tracer.py [hunk 11]
```diff
@@ -61,6 +61,11 @@ def get_tracer(self, tracer_data: TracerData) -> Optional[Tracer]:
 class Tracer(MaintainedModel, ElementLabel):
     objects = TracerQuerySet().as_manager()
 
+    COMPOUND_DELIMITER = "-"
+    LABELS_DELIMITER = ","
+    LABELS_LEFT_BRACKET = "["
+    LABELS_RIGHT_BRACKET = "]"
```
Comment (hepcat72) [inline DataRepo/models/tracer.py:67]: C. Added these.  They did not exist before.

- [ ] TODO: Create an issue to consolidate tThese and the other decorating variables in the other classeswith the values stored independently in the parsing code.

File: DataRepo/models/tracer_label.py [hunk 12]
```diff
@@ -27,6 +27,8 @@ def create_tracer_label(self, tracer: Tracer, isotope_data: IsotopeData):
 class TracerLabel(MaintainedModel, ElementLabel):
     objects = TracerLabelQuerySet().as_manager()
 
+    POSITIONS_DELIMITER = ","
```
Comment (hepcat72) [inline DataRepo/models/tracer_label.py:30]: C. Same here as the file above.

File: DataRepo/models/tracer_label.py [hunk 13]
```diff
@@ -98,10 +100,13 @@ def __str__(self):
...
     )
     def _name(self):
-        # format: `position,position,... - weight element count` (but no spaces) positions optional
+        # format: `position,position,...-MassNumberElementCount`, e.g. 1,2,3-13C3, positions optional (e.g. 13C3)
         positions_string = ""
         if self.positions and len(self.positions) > 0:
-            positions_string = ",".join([str(p) for p in sorted(self.positions)]) + "-"
+            positions_string = (
+                self.POSITIONS_DELIMITER.join([str(p) for p in sorted(self.positions)])
+                + "-"
```
Comment (hepcat72) [inline DataRepo/models/tracer_label.py:108]: C.

- [ ] TODO: Store the dash in a class attribute.

File: DataRepo/tests/loaders/test_tracers_loader.py [hunk 14]
```diff
@@ -68,21 +68,21 @@ def test_init_load(self):
 
         tl.init_load()
 
-        self.assertTrue(hasattr(tl, "tracer_dict"))
+        self.assertTrue(hasattr(tl, "tracers_dict"))
```
Comment (hepcat72) [inline DataRepo/tests/loaders/test_tracers_loader.py:71]: C. This file just has name changes.

File: DataRepo/tests/management/test_load_tracers.py [hunk 15]
```diff
@@ -146,7 +146,7 @@ def test_name_with_multiple_numbers_error(self):
         # tracer name (aspartame-[13C4]): [aspartame]: column [Compound Name] on row [7] in file
         # [DataRepo/data/tests/tracers/tracers_with_errors.tsv]
         self.assertIn("[aspartate]", str(aes.exceptions[3]))
-        self.assertIn("tracer name (aspartame-[13C4])", str(aes.exceptions[3]))
+        self.assertIn("Tracer Name (aspartame-[13C4])", str(aes.exceptions[3]))
```
Comment (hepcat72) [inline DataRepo/tests/management/test_load_tracers.py:149]: C. Not sure how this passed before.  Maybe I changed it...

File: DataRepo/tests/utils/test_infusate_parser.py [hunk 16]
```diff
@@ -6,7 +6,7 @@
 from DataRepo.utils.infusate_name_parser import (
     InfusateData,
     InfusateParsingError,
-    InfusateTracer,
+    InfusateTracerData,
```
Comment (hepcat72) [inline DataRepo/tests/utils/test_infusate_parser.py:9]: C. Had to change this to not collide with the model class of the same name.

File: DataRepo/tests/utils/test_infusate_parser.py [hunk 17]
```diff
@@ -231,24 +231,12 @@ def test_malformed_infusate_parsing_with_new_line(self):
...
         name = "lysine-[13C5]\nlysine-[19O2]"
         with self.assertRaisesRegex(TracerParsingError, "cannot be parsed"):
             _ = parse_tracer_string(name)
 
-    def test_malformed_tracer_parsing_with_improper_delimiter(self):
-        """Test bad tracer delimiter (',' instead of ';')"""
-        name = "lysine-[13C5],glucose-[19O2]"
-        with self.assertRaisesRegex(TracerParsingError, "cannot be parsed"):
-            _ = parse_tracer_string(name)
-
```
Comment (hepcat72) [inline DataRepo/tests/utils/test_infusate_parser.py:251]: C. These 2 tests would no longer raise an exception because of the bug fix to the infusate parsing code to handle an edge case.  The compound pattern is permissing, so these cases would result in an exception elsewhere anyway.

File: DataRepo/utils/exceptions.py [hunk 18]
```diff
@@ -2223,6 +2223,14 @@ def __init__(self, name, **kwargs):
         self.name = name
 
 
+class RecordDoesNotExist(InfileError, ObjectDoesNotExist):
+    def __init__(self, model, name, **kwargs):
+        message = f"{model.__name__} [{name}] from %s does not exist."
+        super().__init__(message, **kwargs)
+        self.model = model
+        self.name = name
```
Comment (hepcat72) [inline DataRepo/utils/exceptions.py:2231]: C. New exception class to provide file context to non-existing record errors.

File: DataRepo/tests/utils/test_infusate_parser.py (hunk 17 above)
Comment (hepcat72) [reply DataRepo/tests/utils/test_infusate_parser.py:251]: See the comment in the infusate parser below.

File: DataRepo/utils/infusate_name_parser.py [hunk 19]
```diff
@@ -16,9 +16,9 @@
 )
 ISOTOPE_ENCODING_JOIN = ","
 ISOTOPE_ENCODING_PATTERN = re.compile(
-    r"(?P<all>(?:(?P<positions>[0-9,]+)-)?(?P<mass_number>[0-9]+)(?P<element>["
+    r"(?P<all>(?:(?P<positions>[0-9][0-9,]*)-)?(?P<mass_number>[0-9]+)(?P<element>["
     + KNOWN_ISOTOPES
-    + r"]{1,2})(?P<count>[0-9]+))"
+    + r"]{1,2})(?P<count>[0-9]+))($|,)"
```
Comment (hepcat72) [inline DataRepo/utils/infusate_name_parser.py:21]: C. Note, this bug (fixed here) is described in #911

File: DataRepo/tests/loaders/test_infusates_loader.py [hunk 20]
```diff
@@ -0,0 +1,439 @@
...
+        self.assertTrue(hasattr(tl, "inconsistent_group_names"))
+        self.assertTrue(hasattr(tl, "inconsistent_names"))
+        self.assertTrue(hasattr(tl, "inconsistent_numbers"))
+
+        self.assertEqual(defaultdict, type(tl.infusates_dict))
+        self.assertEqual(defaultdict, type(tl.infusate_name_to_number))
+        self.assertEqual(dict, type(tl.valid_infusates))
+        self.assertEqual(defaultdict, type(tl.inconsistent_group_names))
+        self.assertEqual(defaultdict, type(tl.inconsistent_names))
+        self.assertEqual(defaultdict, type(tl.inconsistent_numbers))
```
Comment (hepcat72) [inline DataRepo/tests/loaders/test_infusates_loader.py:96]: C. I agree with the feedback in a past PR WRT the corresponding tracers loader tests...

- [ ] TODO: Delete these assertions

File: DataRepo/tests/loaders/test_infusates_loader.py [hunk 21]
```diff
@@ -0,0 +1,439 @@
...
+
+        self.assertEqual(0, len(tl.infusates_dict.keys()))
+        self.assertEqual(0, len(tl.infusate_name_to_number.keys()))
+        self.assertEqual(0, len(tl.valid_infusates.keys()))
+        self.assertEqual(0, len(tl.inconsistent_group_names.keys()))
+        self.assertEqual(0, len(tl.inconsistent_names.keys()))
+        self.assertEqual(0, len(tl.inconsistent_numbers.keys()))
+
+    def test_infusate_loader_load_data(self):
```
Comment (hepcat72) [inline DataRepo/tests/loaders/test_infusates_loader.py:105]: C. From this point onward, these tests are nearly a verbatim copy of the tracers loader tests.

**Comment by hepcat72** [comment]
A rebase went awry and now I seem to be missing work I did yesterday?! Ugh. Looks like I have to redo it.

**Comment by hepcat72** [comment]
Well, I'm going to hope that the work I thought I remembered doing yesterday at home is actually work I did Friday in the office and that it still exists there.  For some reason, I cannot connect remotely to my iMac at work, which I normally can connect to.  I **really** would like to know if that work is there because I don't want to have to do that again.  The more I think about it, the more I think I did it in the office on Friday, because I remember thinking it'd taken all day and I definitely only spent a couple hours tinkering with code yesterday.

**Comment by hepcat72** [comment]
Ah! I found some of the changes using a function I'd created to grep all code ever committed to the repo:
```
git grep "$@" `git rev-list --all`
```
I thought of a unique string to grep for in the recent code and I found a few commits containing it.

But that commit is not in the git log of main. I'm pretty sure they were changes made in the tracers loader branch that was merged.  What a mess.  I hate rebasing.  I have to figure out how to recover this work.

**Comment by hepcat72** [comment]
Ah.  Phew! I found it.  And I learned a new command: `git branch --contains <commit>`.  It tells you what branch(es) contain(s) a commit hash.  The changes are safe in the `load_infusates` branch (the branch off this branch).  The thing that had confused me was I had merged a couple unrelated branches into main and when I went to rebase this branch (which I'd already rebased yesterday based on other changes merged into main from unrelated branches), I found myself fixing things I'd already fixed and I got confused, thinking I'd lost work. It's going to make for a sticky merge between this branch and the `load_infusates` branch, but at least I didn't lose work.

**Comment by hepcat72** [comment]
As I wrap up the peak annotations loader (getting quite close [in terms of the functionality - I still have to break up the imminent PR]), I'm getting close to the point where I will need this merged in order to polish off the submission process refactor.  I know there have been some differences in opinion over the 2 ways to create infusates and tracers in the resulting excel sheets (either via the granular columns or via pasting the names into the name column for parsing by the loading code), but I would say that that is something we can tweak later (to which I proposed I could make the name column read-only).  Either way, the sheet will be auto-populated using existing tracers and infusates based on the contents of the peak annotations files used in the build a submission page.

**Comment by hepcat72** [comment]
Merged via #1076
